File: src/main/kotlin/ru/exrates/entities/exchanges/BasicExchange.java
package ru.exrates.entities.exchanges;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.hibernate.annotations.SortComparator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.configurationprocessor.json.JSONException;
import org.springframework.http.HttpStatus;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;
import ru.exrates.entities.CurrencyPair;
import ru.exrates.entities.TimePeriod;
import ru.exrates.entities.exchanges.secondary.*;
import ru.exrates.utils.JsonSerializers;

import javax.annotation.PostConstruct;
import javax.persistence.*;
import java.time.Duration;
import java.time.Instant;
import java.util.*;

@Entity @Inheritance(strategy = InheritanceType.SINGLE_TABLE) @DiscriminatorColumn(name = "EXCHANGE_TYPE")
@JsonIgnoreProperties({"id", "limits", "limitcode", "banCode", "sleepValueSeconds", "updatePeriod", "temporary"})
public abstract class BasicExchange implements Exchange {

    @Id @GeneratedValue

    private Integer id;

    boolean temporary = true;
    private final static Logger logger = LogManager.getLogger(BasicExchange.class);

    static String URL_ENDPOINT, URL_CURRENT_AVG_PRICE, URL_INFO, URL_PRICE_CHANGE, URL_PING, URL_ORDER;

    @ManyToMany(cascade = CascadeType.PERSIST, fetch = FetchType.EAGER)

    //@JsonIgnore //TODO delete
    @JsonSerialize(using = JsonSerializers.TimePeriodListSerializer.class)
    List<TimePeriod> changePeriods;

    @OneToMany(orphanRemoval = true, cascade = CascadeType.PERSIST, fetch = FetchType.EAGER)

    Set<Limit> limits;

    int limitCode, banCode, sleepValueSeconds = 30;

    Duration updatePeriod;

    @Getter
    String name;


    @OneToMany(cascade = {CascadeType.PERSIST}, fetch = FetchType.EAGER)
    @Getter
    @SortComparator(CurrencyPair.CurComparator.class)
    //@org.hibernate.annotations.OrderBy(clause = "last_use desc")
    //@Column(name="last_use")
    //@OrderColumn(name = "last_use", nullable = false)
    protected final SortedSet<CurrencyPair> pairs = new TreeSet<>();

    @Transient
    private Properties props;

    @Transient
    RestTemplateImpl restTemplate;

    @Transient
    WebClient webClient;

    
    @Autowired
    public void setRestTemplate(RestTemplateImpl restTemplate) {
        this.restTemplate = restTemplate;
    }


    @Autowired
    public void setProps(Properties props) {
        this.props = props;
    }



    BasicExchange() {
        // init pairs

    }

    @PostConstruct
    protected void init(){
        logger.debug("Postconstruct " + name);
        TimerTask task = new TimerTask() {
            @Override
            public void run() {
                try {
                    task();
                }catch (RuntimeException e){
                    this.cancel();
                    logger.error(e);
                    logger.debug(e.getMessage());
                }
            }
        };
        Timer timer = new Timer();
        updatePeriod = Duration.ofMillis(props.getTimerPeriod());
        timer.schedule(task, 10000, props.getTimerPeriod());
        //timer.cancel();
    }

    @Override
    public void insertPair(CurrencyPair pair){
        pairs.add(pair);
        if (pairs.size() > props.getMaxSize()) pairs.remove(pairs.first());
    }

    @Override
    public CurrencyPair getPair(Currency c1, Currency c2){
        final CurrencyPair[] pair = {null};
        pairs.spliterator().forEachRemaining(el ->  {
            if (el.getSymbol().equals(c1.getSymbol() + c2.getSymbol())) pair[0] = el;
        });
        return pair[0];
    }

    @Override
    public CurrencyPair getPair(String pairName){
        final CurrencyPair[] p = {null};
        pairs.spliterator().forEachRemaining(el ->  {
            if (el.getSymbol().equals(pairName)) p[0] = el;
        });
        return p[0];

    }

    public boolean dataElapsed(CurrencyPair pair, Duration timeout, int idx){
        logger.debug(String.format("Pair %1$s was updated on field %2$s %3$s | current time = %4$s", pair.getSymbol(), idx, Instant.ofEpochMilli(pair.getUpdateTimes()[idx]), Instant.now()));
        return Instant.now().isAfter(Instant.ofEpochMilli(pair.getUpdateTimes()[idx] + timeout.toMillis()));
    }


    protected void task() throws RuntimeException{
        logger.debug( name + " task started....");
        CurrencyPair pair = null;
        synchronized (pairs){
            for (CurrencyPair p : pairs) {
                try {
                    //p.getPriceHistory().add(3.2);
                    //p.getPriceHistory().add(3.5);
                    currentPrice(p, updatePeriod);
                    priceChange(p, updatePeriod);
                } catch (JSONException e) {
                    logger.error("task JS ex", e);
                } catch (LimitExceededException e){
                    logger.error(e.getMessage());
                    try {
                        sleepValueSeconds *= 2;
                        Thread.sleep(sleepValueSeconds);
                    } catch (InterruptedException ex) {
                        logger.error("Interrupt ", ex);
                    }
                    task();
                    return;
                } catch (ErrorCodeException e){
                    logger.error(e.getMessage());
                } catch (BanException e){
                    logger.error(e.getMessage());
                    throw new RuntimeException("You are banned from " + this.name);
                } catch (Exception e){
                    throw new RuntimeException("Unknown error", e);
                }
            }
        }

    };

    <T> T request(String uri, Class<T> tClass) throws BanException, LimitExceededException, IllegalStateException{
        return webClient.get().uri(uri).retrieve().onStatus(HttpStatus::is4xxClientError, resp ->{
            Exception ex = null;
            switch(resp.statusCode().value()){
                case 418: ex = new BanException();
                    break;
                case 429: ex = new LimitExceededException(LimitType.WEIGHT);
                    break;
                default:
                    throw new IllegalStateException("Unexpected value: " + resp.statusCode().value());
            }
            return Mono.error(ex);
        }).bodyToMono(tClass).block();
    }

    public abstract void currentPrice(CurrencyPair pair, Duration timeout) throws
            JSONException, NullPointerException, LimitExceededException, ErrorCodeException, BanException;

    public abstract void priceChange(CurrencyPair pair, Duration timeout, Map<String, String> uriVariables) throws
            JSONException, LimitExceededException, ErrorCodeException, BanException;

    public abstract void priceChange(CurrencyPair pair, Duration timeout) throws
            JSONException, LimitExceededException, ErrorCodeException, BanException;



    @Override
    public String toString() {
        return "BasicExchange{" +
                "id=" + id +
                ", temporary=" + temporary +
                ", changePeriods=" + changePeriods +
                ", limits=" + limits +
                ", limitCode=" + limitCode +
                ", banCode=" + banCode +
                ", sleepValueSeconds=" + sleepValueSeconds +
                ", updatePeriod=" + updatePeriod +
                ", name='" + name + '\'' +
                ", pairs=" + pairs +
                ", props=" + props +
                ", restTemplate=" + restTemplate +
                '}';
    }
}

File: src/main/kotlin/ru/exrates/entities/exchanges/BasicExchange.kt
package ru.exrates.entities.exchanges

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.apache.logging.log4j.Logger
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode
import org.hibernate.annotations.SortComparator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.ApplicationContext
import ru.exrates.configs.Properties
import ru.exrates.entities.CurrencyPair
import ru.exrates.entities.TimePeriod
import ru.exrates.entities.exchanges.rest.HuobiExchange
import ru.exrates.entities.exchanges.rest.RestExchange
import ru.exrates.entities.exchanges.secondary.BanException
import ru.exrates.entities.exchanges.secondary.ErrorCodeException
import ru.exrates.entities.exchanges.secondary.Limit
import ru.exrates.entities.exchanges.secondary.LimitExceededException
import ru.exrates.func.EndpointStateChecker
import ru.exrates.func.TaskHandler
import ru.exrates.utils.ClientCodes
import ru.exrates.utils.TimePeriodListSerializer
import java.net.ConnectException
import java.time.Duration
import java.util.*
import javax.annotation.PostConstruct
import javax.persistence.*
import kotlin.collections.ArrayList
import kotlin.collections.HashSet

@Entity @Inheritance(strategy = InheritanceType.SINGLE_TABLE) @DiscriminatorColumn(name = "EXCHANGE_TYPE")
@JsonIgnoreProperties("id", "limits", "limitCode", "banCode", "sleepValueSeconds", "taskTimeOut", "temporary",
    "props", "delimiter")
abstract class BasicExchange() : Exchange, Cloneable{
    var exId: Int = 0
    var temporary = true
    var limitCode: Int = 0
    var banCode: Int = 0
    var serverError: Int = 503
    var sleepValueSeconds = 30L

    @Autowired
    @javax.persistence.Transient
    lateinit var props: Properties

    @javax.persistence.Transient
    @Autowired
    lateinit var stateChecker: EndpointStateChecker

    @Autowired
    @javax.persistence.Transient
    lateinit var applicationContext: ApplicationContext

    @javax.persistence.Transient
    @Autowired
    lateinit var logger: Logger

    @Autowired
    @javax.persistence.Transient
    lateinit var taskHandler: TaskHandler

    @Id @GeneratedValue
    var id: Int = 0
    var name: String = ""
    var delimiter = ""

    @OneToMany(cascade = [CascadeType.PERSIST], fetch = FetchType.EAGER)
    @SortComparator(CurrencyPair.SortComparator::class)
    val pairs: SortedSet<CurrencyPair> = TreeSet<CurrencyPair>()

    @ManyToMany(cascade = [CascadeType.PERSIST], fetch = FetchType.EAGER)
    @Fetch(FetchMode.SELECT)
    @JsonSerialize(using = TimePeriodListSerializer::class)
    val changePeriods: MutableList<TimePeriod> = ArrayList()

    @OneToMany(orphanRemoval = true, cascade = [CascadeType.PERSIST], fetch = FetchType.EAGER)
    val limits: MutableSet<Limit> = HashSet()

    @OneToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH])
    var taskTimeOut: TimePeriod = TimePeriod(Duration.ofMillis(90000), "1.5m")

    @ElementCollection(fetch = FetchType.EAGER)
    @Fetch(FetchMode.SELECT)
    lateinit var historyPeriods: List<String>

    @ElementCollection(fetch = FetchType.EAGER)
    val topPairs: MutableList<String> = LinkedList()


    /*
    * ******************************************************************************************************************
    *       Initialization
    * ******************************************************************************************************************
    * */

    @PostConstruct
    fun init(){
        logger.debug("Postconstruct super $name")
        changePeriods.sortBy { it.period }
        if (!props.taskIsEnabled()) return
        GlobalScope.launch {
            launch{
                delay(10000)
                repeat(Int.MAX_VALUE){
                    task()
                    logger.debug("Task started with ${taskTimeOut.period.toMillis()} millis")
                    delay(taskTimeOut.period.toMillis())
                }
            }
        }
    }

    fun task(){
        logger.debug("$name task started with ${pairs.size} pairs")
        logger.debug("pairs in exchange: ${pairs.joinToString { it.symbol }}")
            for (p in pairs){
                try {
                    with(taskHandler){
                        logger.debug("Updating task data for $p in $name")
                        runTasks((this@BasicExchange as RestExchange).requestDelay(),
                            { currentPrice(p, taskTimeOut) },
                            { priceChange(p, taskTimeOut) },
                            { priceHistory(p, changePeriods[0].name, 10) } //todo [0] right?
                        )
                    }

                }catch (e: LimitExceededException){
                    logger.error(e.message)
                    sleepValueSeconds *= 2
                    Thread.sleep(sleepValueSeconds)
                    task()
                    return
                }catch (e: ErrorCodeException){ logger.error(e.message)
                }catch (e: BanException){
                    logger.error(e.message)
                    throw RuntimeException("You are banned from $name")
                }catch (e: ConnectException){
                    logger.error(e)
                }
            }

    }



    /*
    * ******************************************************************************************************************
    *       Update methods
    * ******************************************************************************************************************
    * */

    abstract fun currentPrice(pair: CurrencyPair, period: TimePeriod)

    abstract fun priceChange(pair: CurrencyPair, interval: TimePeriod)

    fun priceHistory(pair: CurrencyPair, interval: String, limit: Int){
        if (!this.historyPeriods.contains(interval)) {
            logger.error("History period $interval incorrect for ${this.name} exchange")
            return
        }
    }


    /*
    * ******************************************************************************************************************
    *       Class methods
    * ******************************************************************************************************************
    * */

    override fun insertPair(pair: CurrencyPair) {
        pairs.add(pair)
        if(pairs.size > props.maxSize()) {
            pairs.last().exchange = EmptyExchange()
            pairs.remove(pairs.last())
        }
    }

    override fun getPair(c1: String, c2: String): CurrencyPair? = pairs.find{c1 == it.baseCurrency && c2 == it.quoteCurrency}

    override fun getPair(pairName: String): CurrencyPair? = pairs.find { it.symbol == pairName }

    fun getTimePeriod(period: String) = changePeriods.find { it.name == period } ?: throw NullPointerException("period in ${this.name} for $period not found")

    fun getTimePeriod(duration: Duration) = changePeriods.find { it.period == duration } ?: throw NullPointerException("period in ${this.name} for $duration not found")

    fun castToRestExchange() = this as RestExchange

}

class ExchangeDTO(exchange: BasicExchange?, exName: String = "", stat: Int = ClientCodes.SUCCESS){
    val exId= exchange?.exId ?: 0
    val name= exchange?.name ?: exName
    @JsonSerialize(using = TimePeriodListSerializer::class)
    val changePeriods= exchange?.changePeriods ?: listOf<TimePeriod>()
    val historyPeriods= exchange?.historyPeriods ?: emptyList()
    var pairs: SortedSet<CurrencyPair> = TreeSet<CurrencyPair>(exchange?.pairs?.filter { it.status == ClientCodes.SUCCESS } ?: TreeSet<CurrencyPair>())
    var status: Int = stat
    init {
       if(exchange == null) this.status = ClientCodes.EXCHANGE_NOT_FOUND
    }

    override fun toString(): String {
        return "\n$name exId = $exId pairs: ${pairs.joinToString { it.symbol }}"
    }

}